docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,136 @@
+
+```
+(RANDLIST) \forall Is:List{Int} . \forall I:Int . \neg(I \in Is) -> randDifferentFromList(Is) -> * I
+```
+
+### Frontend Syntax
+
+We do not have a frontend syntax proposal for `rand()` and its variants. This is left as future work. 
+
+## Example B: Fresh Integer Construct `fresh(Is)`
+
+The fresh integer construct `fresh(Is)` is a language construct and part of the PL syntax
+
+```
+Stmt ::= ... | "fresh" "(" List{Int} ")"
+```
+
+### Axioms
+
+The intended semantics of `fresh(Is)` is that it can rewrite to any integers that are not in `Is`.  
+
+Note that `fresh(Is)` and `randDifferentFromList(Is)` are different; the former does not need to rewrite to _every_ integers not in `Is`, while the latter requires so. For example, it is _correct_ to implement `fresh(Is)` so it always returns the smallest positive integers that are not in `Is`, but same implementation for `randDifferentFromList(Is)` is considered _incorrect_. In other words, there exist multiple correct implementations of `fresh(Is)`, some of which may be deterministic, but there only exists a unique implementation of `randDifferentFromList(Is)`. Finally, it is a _correct_ if one implements `fresh(Is)` the same as `randDifferentFromList(Is)`. 
+
+Here is the axiom of `fresh(Is)`
+
+```
+(FRESH) \forall Is:List{Int} . \exists I:Int . \neg( I \in Is) /\ (fresh(Is) -> * I)
+// equivalent form(s)
+//why?  \forall Is:List{Int} . fresh(Is) -> \exists I:Int . \neg( I \in Is) /\ * I
+//      \forall Is:List{Int} . fresh(Is) -> \exists I:Int . * (\neg( I \in Is) /\ I)
+```
+
+### Frontend Syntax
+
+We use the following K syntax to define `fresh(Is)`
+
+```
+rule fresh(Is:List{Int}) => ?I:Int ensures notBool(?I in Is)
+```
+
+## Example C: Arbitrary Number (Unspecific Function) `arb()`
+
+The function `abr()` is not a PL construct, but a mathematical function. Therefore, its axioms do not use rewrites but equalities. 
+
+### Axioms
+
+The intended semantics of `arb()` is that it is an unspecified nullary function. The exact return value of `arb()` is unspecified in the semantics but up to the implementations. However, being a mathematical function, `arb()` must return the same value in any one implementation. 
+
+The axiom of `arb()` is 
+
+```
+(ARB) \exists I:Int . arb() = I
+//    this is exactly (Function) axiom
+//    this is not equivalent to
+//      arv() = (\exists I:Int . I)
+```
+
+There are many variants of `arb()`. For example. `arbDifferentFromList(Is)` is an unspecific function whose return value must be different than those in `Is`. Here is its axiom
+
+```
+(ARBDIFF) \forall Is:List{Is} . \exists I:Int . \neg(I \in Is) /\ arbDifferentFromList(Is) = I
+```
+
+### Frontend Syntax
+
+We do not need special frontend syntax to define `arb()`. We only need to define it in the usual way as a function (instead of a language construct). 
```

Comment:
if we treat `arb()` as uninterpreted function, the current frontend syntax is enough. However, if we want to rewrite it to some variable, the current frontend syntax doesn't support it.